### PR TITLE
Snapshots: allow snapshotting some user code

### PR DESCRIPTION
**Pull request by hoodmane**
This adds some basic ability to snapshot after executing user code. It is pretty brittle right now:
1. It will crash if the user loads any binary extensions before taking the snapshot
2. It doesn't track changes to the file system
3. It doesn't mind there being non-reproducible objects in the hiwire map

Snapshots will probably have to be experimental for quite a while. This should ideally get a bunch of tests.

1. I think I have a pretty good solution for this, which I can work on in a followup.

2. One possibility here is we could serialize the entire filesystem state into the memory snapshot. This would be hard and make the snapshot big, but we wouldn't have to load python_stdlib.zip when restoring from a snapshot so it probably wouldn't increase the total download size by much...

3. Should be pretty easy to fix.

cc @pfebrer

### Checklists

- [ ] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
- [x] Add / update tests
- [ ] Add new / update outdated documentation

### CODE REVIEW

**Comment by pfebrer** [comment]
Thanks for the heads up! Regarding point 1, does that mean that only pure python packages can be imported before generating the snapshot?

**Comment by hoodmane** [comment]
Yeah after this PR that is still the case. I will remove this restriction in the next PR. To fix it we have to patch this line in Emscripten:
https://github.com/emscripten-core/emscripten/blob/main/src/library_dylink.js#L633
When taking the snapshot we need to store the pointer to the dylink metadata for each loaded .so file into our snapshot metadata. When restoring the snapshot, we need to make sure that we put the dylink metadata at the same address as before. We also have to store all the handles returned from `dlopen` that haven't been `dlclosed`.

**Comment by pfebrer** [comment]
Cool! To test it in my application I need to import packages with C extensions. Point 2 is not a problem and I don't understand what point 3 means 😅

Happy to see that you are getting closer to it!

**Comment by hoodmane** [comment]
For number 3, the issue is if you did something like:
```
pyodide.runPython(`
from js import fetch
some_var = fetch(some_url)
`);
```
Pyodide says okay `some_var` is a `JsProxy` with table index 10. But the entry in table index 10 is a `Promise<FetchResponse>` which is not something we can really serialize. So if you try to make a snapshot after executing this code, it should throw an error explaining this. Unfortunately in real code it is be pretty hard for us to trace through and explain why the `JsProxy` exists but we can at least say what the value is.

**Comment by hoodmane** [comment]
I guess we'll also have a problem if ctypes is used...

File: src/js/snapshot.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
+import { jsFinderHook } from "./api";
+import { scheduleCallback } from "./scheduler";
+
+declare var Module: any;
+const MAP_INDEX = 5;
```
Comment (ryanking13) [inline src/js/snapshot.ts:5]: Where does this magic number comes from?

File: src/tests/test_snapshots.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                from js import Headers, URL
+
+                a = Headers.new()
+            `);
+            py1.makeMemorySnapshot();
+            """
+        )
+
+
+def test_snapshot_deleted_proxy(selenium_standalone_noload):
```
Comment (ryanking13) [inline src/tests/test_snapshots.py:60]: What does this test trying to test?

**Review by ryanking13 — APPROVED** [review]
Thanks. To be honest, I don't understand the code inside `snapshot.ts`, but it somehow works so I'll approve so you can do further experiments :) I hope we can clean up the code when this feature gets more stable.

File: packages/gensim/meta.yaml [hunk 3]
```diff
@@ -16,6 +16,7 @@ requirements:
     - scipy
     - six
     - smart_open
+    - wrapt
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply packages/gensim/meta.yaml:19]: Indeed. Oops.

File: src/templates/makesnap.mjs [hunk 4]
```diff
@@ -1,5 +1,11 @@
 import { loadPyodide } from "./pyodide.mjs";
 import { writeFileSync } from "fs";
+import { fileURLToPath } from "url";
+import { dirname } from "path";
+
+const __dirname = dirname(fileURLToPath(import.meta.url));
 
 const py = await loadPyodide({ _makeSnapshot: true });
-writeFileSync("snapshot.bin", py._snapshot);
+py.makeMemorySnapshot();
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply src/templates/makesnap.mjs:9]: It's a mistake. Thanks.

File: packages/gensim/meta.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryanking13) [reply packages/gensim/meta.yaml]: My git skills haven't improved in years too haha. I still do `git checkout` to make branches.

[the remark this answers is not in the document]
Comment (hoodmane) [reply packages/gensim/meta.yaml]: Yeah me too. Is there another way?

[the remark this answers is not in the document]
Comment (hoodmane) [reply packages/gensim/meta.yaml]: Ah, I use `git restore` from that group of commands a lot but I never learned to use `git switch`.

**Pull request merged**